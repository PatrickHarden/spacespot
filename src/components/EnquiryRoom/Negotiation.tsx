import React from 'react'
import { get } from 'lodash'
import { useIntl } from 'react-intl'
import { useSelector } from 'react-redux'
import styled from 'styled-components'
import device from 'services/global/device'
import userSelectors from 'services/user/selectors'
import negotiationSelectors from 'services/negotiation/selectors'
import enquirySelectors from 'services/enquiry/selectors'
import { NegotiationStatus } from 'services/negotiation/types'
import { useSigners } from 'services/negotiation/hooks'
import LeaseTerms from './LeaseTerms'
import SelectFitOut from './SelectFitOut'
import SpecialProvisions from './SpecialProvisions'
import Summary from './Summary'
import SummaryMobile from './SummaryMobile'
import Heading2 from 'components/common/Heading2'
import Congrats from './Congrats'
import spaceSelector from 'services/space/selectors'

const Explanation = styled.div`
  color: #000000;
  font-size: 16px;
  font-weight: 400;
`
const Block = styled.div`
  h1 {
    color: #000000;
    font-size: 25px;
    font-weight: 500;
    letter-spacing: -0.23px;
    line-height: 33px;
  }
`
const ContainerWrapper = styled.div`
  max-width: 1260px;
  margin: 0 auto;
`

const Container = styled.div`
  padding: 10px 20px;
  margin: 0 24px;
  display: flex;
  flex-direction: row;
  @media ${device.mobile} {
    flex-direction: column;
    padding: 10px 10px;
    margin: 0 0 100px 0;
  }
`
const Col1 = styled.div<{ disabled: boolean }>`
  flex: 1 0 70%;
  padding-top: 70px;
  @media ${device.mobile} {
    ${props => (props.disabled ? 'display: none;' : 'flex: 1 0 100%;')}
  }
`

const Col2 = styled.div<{ disabled: boolean }>`
  flex: 0 1 30%;
  margin-left: 20px;
  @media ${device.mobile} {
    margin: ${props => (props.disabled ? '0 auto' : '0 0 0 20px')};
    ${props => (props.disabled ? 'flex: 1 0 100%;' : 'display: none;')}
  }
`

const MobileFooter = styled.div<{ disabled: boolean }>`
  display: none;
  ${props =>
    props.disabled
      ? ''
      : `
  @media ${device.mobile} {
    z-index: 2;
    display: block;
    position: fixed;
    left: 0;
    bottom: 0;
    width: 100%;
    background: white;
  }`}
`
const Heading2MarginTop = styled(Heading2)`
  margin-top: 70px;
`
const Negotiation = () => {
  const { formatMessage } = useIntl()
  const t = (s: string) => formatMessage({ id: s })
  const enquiry = useSelector(enquirySelectors.getEnquiry)
  const isLandlord = useSelector(userSelectors.isLandlord)
  const selectedProvision = useSelector(
    negotiationSelectors.getNegotiationSpecialProvisions,
  )
  const negotiationStatus = useSelector(
    negotiationSelectors.getNegotiationStatus,
  )
  const fitoutOptions = useSelector(spaceSelector.selectSpaceFitout)

  const signers = useSigners()
  if (signers) {
    const pending = signers.filter(d => d.signerStatus !== 'completed')
    const bothSigned = signers.length > 0 && pending.length === 0
    if (bothSigned) {
      return <Congrats />
    }
  }

  const disabled =
    negotiationStatus !== NegotiationStatus.TermsAgreed &&
    negotiationStatus !== NegotiationStatus.TermsNotAgreed
  const isFlex = get(enquiry, 'flex', false)
  const foTitle = isLandlord
    ? t('NEGOTIATE_FIT_OUT_LANDLORD')
    : fitoutOptions && fitoutOptions.length > 1
    ? t('NEGOTIATE_FIT_OUT_TENANT_SELECTED')
    : t('NEGOTIATE_FIT_OUT_TENANT')

  const spTitle =
    !isLandlord && selectedProvision !== ''
      ? t('NEGOTIATE_PROVISIONS_TENANT_ACCEPT')
      : t('NEGOTIATE_PROVISIONS')
  return (
    <ContainerWrapper>
      <Container>
        <Col1 disabled={disabled}>
          <Block>
            <Explanation>
              {isLandlord
                ? t('NEGOTIATE_LANDLORD_EXPLANATION_1')
                : t('NEGOTIATE_EXPLANATION_1')}
            </Explanation>
            <Explanation>
              {isLandlord
                ? t('NEGOTIATE_LANDLORD_EXPLANATION_2')
                : t('NEGOTIATE_EXPLANATION_2')}
            </Explanation>
            <Explanation>
              {isLandlord
                ? t('NEGOTIATE_LANDLORD_EXPLANATION_3')
                : t('NEGOTIATE_EXPLANATION_3')}
            </Explanation>
          </Block>
          <Block>
            <Heading2MarginTop>{spTitle}</Heading2MarginTop>
            <SpecialProvisions disabled={disabled} />
          </Block>
          <Block>
            <Heading2>{t('NEGOTIATE_LEASE_TERMS')}</Heading2>
            <LeaseTerms disabled={disabled} />
          </Block>
          {!isFlex && (
            <Block>
              <Heading2MarginTop>{foTitle}</Heading2MarginTop>
              <SelectFitOut disabled={disabled} />
            </Block>
          )}
        </Col1>
        <Col2 disabled={disabled}>
          <Summary />
        </Col2>
        <MobileFooter disabled={disabled} id="footer">
          <SummaryMobile />
        </MobileFooter>
      </Container>
    </ContainerWrapper>
  )
}

export default Negotiation
