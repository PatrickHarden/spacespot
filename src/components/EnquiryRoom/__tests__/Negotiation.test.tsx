import React from 'react'
import { render } from '@testing-library/react'
import { createMockStore } from 'redux-test-utils'
import { Provider } from 'react-redux'

import Negotiation from '../Negotiation'

jest.mock('react-intl', () => ({
  useIntl: () => ({ formatMessage: (s: any) => s.id }),
  createIntlCache: () => jest.fn(),
  createIntl: () => ({ formatMessage: (s: any) => s.id }),
}))

const mockDispatch = jest.fn()
const mockState = {
  enquiry: {
    data: {
      spaceId: 'SS_SPACE_665',
      peer1Id: '1',
      peer2Id: '2',
      chatId: 159,
      enquireStatus: 'ACCEPTED',
      negotiationId: 139,
      enquireId: 159,
      flex: false,
    },
  },
  negotiation: {
    data: {
      flexible: false,
      terms: [
        {
          termId: 5,
          value: 'MONTHLY',
        },
        {
          termId: 6,
          value: '2000',
        },
        {
          termId: 7,
          value: '2',
        },
        {
          termId: 8,
          value: '2000',
        },
        {
          termId: 9,
          value: 'GUARANTEE',
        },
        {
          termId: 1,
          value: '3500',
        },
        {
          termId: 2,
          value: '40000',
        },
        {
          termId: 3,
          value: '2020-04-22',
        },
        {
          termId: 4,
          value: '13',
        },
      ],
      negotiationId: 139,
      enquireId: 159,
      createdAt: '2020-02-14T06:30:08.354+0000',
      lastUpdated: '2020-03-05T16:35:02.513+0000',
      lastModifyBy: '0062fb00-e389-4bb8-9119-613e3c45938a',
      status: 'PROPOSED',
      customFitout: false,
      fitoutOption: 0,
      specialProvision: 'gfd',
      specialProvisionStatus: 'CREATED',
    },
    loading: false,
    termsFields: {
      '1': 'Area',
      '2': 'Rent',
      '3': 'Start',
      '4': 'Duration',
      '5': 'Frequency',
      '6': 'AdditionalCosts',
      '7': 'Deposit',
      '8': 'ServiceCharges',
      '9': 'DepositType',
    },
    customFitout: [],
  },
  user: {
    auth: {
      idToken: {
        claims: {
          // eslint-disable-next-line @typescript-eslint/camelcase
          extension_Landlord: true,
        },
      },
    },
  },
}

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}))

it('Renders without crash', () => {
  const store: any = createMockStore(mockState)
  const wrapper = render(
    <Provider store={store}>
      <Negotiation />
    </Provider>,
  )
  expect(wrapper.asFragment()).toMatchSnapshot()
})
