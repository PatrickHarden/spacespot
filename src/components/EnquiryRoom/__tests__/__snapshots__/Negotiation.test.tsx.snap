// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crash 1`] = `
<DocumentFragment>
  <div
    class="sc-eMigcr gKPvBo"
  >
    <div
      class="sc-fzsDOv dyVjQx"
    >
      <div
        class="sc-itybZL iEUZkf"
      >
        <div
          class="sc-dTdPqK dfprlY"
        >
          NEGOTIATE_EXPLANATION_1
        </div>
        <div
          class="sc-dTdPqK dfprlY"
        >
          NEGOTIATE_EXPLANATION_2
        </div>
        <div
          class="sc-dTdPqK dfprlY"
        >
          NEGOTIATE_EXPLANATION_3
        </div>
      </div>
      <div
        class="sc-itybZL iEUZkf"
      >
        <h2
          class="sc-kPVwWT sc-kUaPvJ fwOsuB"
        >
          NEGOTIATE_FIT_OUT_LANDLORD
        </h2>
        <div
          class="sc-gGBfsJ gxAkCg"
        >
          <h3
            class="sc-hqyNC jWkGJW"
          >
            CUSTOM_FITOUT_EMPTY
          </h3>
          <button
            class="sc-brqgnP tnlpq"
          >
            <svg
              class="sc-jnlKLf dZWbHL"
              height="1.1em"
              viewBox="0 0 18 17"
              width="1.1em"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke="#6CB9D5"
                stroke-linecap="round"
                stroke-width="2"
              >
                <path
                  d="M9 1.6v13.8M1.64 8.5h14.72"
                />
              </g>
            </svg>
             CUSTOM_FITOUT_OPTION_ADD
          </button>
        </div>
      </div>
      <div
        class="sc-itybZL iEUZkf"
      >
        <h2
          class="sc-kPVwWT sc-kUaPvJ fwOsuB"
        >
          NEGOTIATE_PROVISIONS
        </h2>
        <div
          class="sc-elJkPf jYklyF"
        >
          <div
            class="sc-ktHwxA gfswcL"
          >
            <p>
              <span>
                SPECIAL_PROV_TEMPLATE_MSG: 
              </span>
              <button
                class="sc-kafWEX kZKReJ"
                data-testid="download"
              >
                SPECIAL_PROV_TEMPLATE_STANDARD
              </button>
            </p>
            <div>
              <div
                class=""
                tabindex="0"
              >
                <input
                  accept="application/pdf"
                  autocomplete="off"
                  multiple=""
                  style="display: none;"
                  tabindex="-1"
                  type="file"
                />
                <button
                  class="sc-cIShpX jNfJhG"
                  data-testid="upload"
                >
                  <img
                    alt="upload"
                    src="upload.svg"
                  />
                  SPECIAL_PROV_TEMPLATE_UPLOAD
                </button>
              </div>
            </div>
          </div>
          <div
            class="sc-kTUwUJ ftkIjf"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root sc-dqBHgY dtXTuM"
              data-testid="removeSP"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <div>
              gfd
            </div>
          </div>
          <div
            class="sc-iyvyFf kDLJQy"
          >
            <button
              class="sc-brqgnP sc-feJyhm ijLrqQ"
            >
              SPECIAL_PROV_MODIFY
            </button>
            <span
              class="sc-gzOgki hzDEgs"
            >
              SPECIAL_PROV_STATUS_CREATED
            </span>
          </div>
        </div>
      </div>
      <div
        class="sc-itybZL iEUZkf"
      >
        <h2
          class="sc-kPVwWT gmbLKa"
        >
          NEGOTIATE_LEASE_TERMS
        </h2>
        <div
          class="sc-bRBYWo gFRiVi"
        >
          <h2
            class="sc-Rmtcm krzlQq"
          >
            NEGOTIATE_YOUR_TERMS_LANDLORD
          </h2>
          <div
            class="sc-EHOje jZQcXK"
          >
            <svg
              class="sc-gqjmRU cugBST"
              height="54"
              viewBox="-2 -2 58 58"
              width="54"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="white"
                fill-rule="evenodd"
                opacity="1"
                stroke="none"
                stroke-width="1"
              >
                <g
                  fill="#ffcdc8"
                  fill-opacity="1"
                  fill-rule="nonzero"
                  stroke="#fe7062"
                  stroke-width="2"
                >
                  <circle
                    cx="27"
                    cy="27"
                    r="27"
                  />
                </g>
                <g
                  fill="#fe7062"
                  fill-rule="nonzero"
                >
                  <text
                    font-size="26"
                    font-weight="400"
                    id="name"
                    text-anchor="middle"
                  >
                    <tspan
                      x="27"
                      y="37"
                    >
                      IH
                    </tspan>
                  </text>
                </g>
              </g>
            </svg>
            <div
              class="sc-bZQynM hnnyjN"
            >
              <h2
                class="sc-htpNat hDHSmr"
              >
                NEGOTIATE_AREA
              </h2>
              <div
                class="sc-ifAKCX dXjDeY"
              >
                3500sq.m.
              </div>
            </div>
            <div
              class="sc-gzVnrw cZMWQi"
            >
              <h2
                class="sc-htpNat hDHSmr"
              >
                NEGOTIATE_LENGTH
              </h2>
              <div
                class="sc-ifAKCX dXjDeY"
              >
                13 NEGOTIATE_MONTHS NEGOTIATE_FROM ONBOARDING_AVAILABILITY_NOW
              </div>
            </div>
            <div
              class="sc-htoDjs hXNBKs"
            >
              <h2
                class="sc-htpNat hDHSmr"
              >
                NEGOTIATE_MONTHLY NEGOTIATE_PAYMENT
              </h2>
              <div
                class="sc-iwsKbI hiqfNw"
              >
                <div
                  class="sc-gZMcBi fYhffy"
                >
                  <div
                    class="sc-ifAKCX dXjDeY"
                  >
                    40000kr
                  </div>
                  <div
                    class="sc-bxivhb etrvNB"
                  >
                    NEGOTIATE_RENT
                  </div>
                </div>
                <div
                  class="sc-gZMcBi fYhffy"
                >
                  <div
                    class="sc-ifAKCX dXjDeY"
                  >
                    2000kr
                  </div>
                  <div
                    class="sc-bxivhb etrvNB"
                  >
                    NEGOTIATE_SERVICE_CHARGES
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-dnqmqq gJJfia"
            >
              <h2
                class="sc-htpNat hDHSmr"
              >
                NEGOTIATE_DEPOSIT
              </h2>
              <div
                class="sc-ifAKCX dXjDeY"
              >
                2 NEGOTIATE_DEPOSIT_IN NEGOTIATE_DEPOSIT_GARANTEE
              </div>
            </div>
          </div>
          <div
            class="sc-gipzik bJCGzT"
          >
            <div>
              <button
                class="sc-brqgnP sc-jDwBTQ jokWaa"
                data-testid="modify"
              >
                NEGOTIATE_MODIFY
              </button>
            </div>
            <div>
              <span
                class="sc-csuQGl dcYdvi"
              >
                NEGOTIATE_STATUS_WAITING_CONFIRMATION
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-gPWkxV hCsoHJ"
    >
      <div
        class="sc-bXGyLb gjEVdO"
      >
        <div
          class="sc-lkqHmb naxll"
        >
          <h2
            class="sc-kPVwWT sc-daURTG bQvNNR"
          >
            SUMMARY_TITLE
          </h2>
          <div
            class="sc-cbkKFq kIqKWP"
          >
            <div
              class="sc-fYiAbW csePZy"
            >
              SUMMARY_RENT
            </div>
            <div
              class="sc-fOKMvo dRHEOq"
            >
              40,000kr
            </div>
          </div>
          <div
            class="sc-cbkKFq kIqKWP"
          >
            <div
              class="sc-fYiAbW csePZy"
            >
              SUMMARY_SERVICE_CHARGES
            </div>
            <div
              class="sc-fOKMvo dRHEOq"
            >
              2,000kr
            </div>
          </div>
          <div
            class="sc-cbkKFq sc-krvtoX hbdGJC"
          >
            <div
              class="sc-dUjcNx blveqi"
            >
              SUMMARY_MONTHLY_TOTAL
            </div>
            <div
              class="sc-gHboQg crupRU"
            >
              42,000kr
            </div>
          </div>
          <div
            class="sc-cbkKFq kIqKWP"
          >
            <div
              class="sc-eilVRo kYKuKc"
            >
              SUMMARY_LEASE_TOTAL_MONTHS
            </div>
            <div
              class="sc-eerKOB dAsPsz"
            >
              546,000kr
            </div>
          </div>
        </div>
        <div
          class="sc-eLExRp vRCSD"
        >
          <div
            class="sc-kfGgVZ kDLxNN"
          >
            SUMMARY_TEXT_LANDLORD_SUBMIT_TERMS
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-jVODtj kGwmLz"
      id="footer"
    >
      <div
        class="sc-bnXvFD csxtcE"
      >
        <div
          class="sc-gFaPwZ bsUfux"
        >
          <div
            class="sc-fhYwyz fORKGZ"
          >
            SUMMARY_MONTHLY_TOTAL
          </div>
          <div
            class="sc-gJWqzi iXMdab"
          >
            42,000kr
          </div>
        </div>
        <div
          class="sc-gFaPwZ bsUfux"
        >
          <div
            class="sc-fhYwyz fORKGZ"
          >
            SUMMARY_LEASE_TOTAL
          </div>
          <div
            class="sc-jzgbtB eINSbh"
          >
            546,000kr
          </div>
        </div>
        <div
          class="sc-kfGgVZ kDLxNN"
        >
          SUMMARY_TEXT_LANDLORD_SUBMIT_TERMS_MOBILE
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
